File: algorithms/[1011]在 D 天内送达包裹的能力/solution.py
# topics = ["数组"]

from typing import List
from math import ceil


class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        w = max(max(weights), ceil(sum(weights) / D))

        # 判断运载能力为 w 时，能否在 D 天内运输完成
        def helper(w: int) -> bool:
            total = day = 0
            for weight in weights:
                if total + weight > w:
                    total = weight
                    day += 1
                else:
                    total += weight
            return day + bool(total) <= D

        while True:
            if helper(w):
                return w
            w += 1

File: algorithms/[1011]在 D 天内送达包裹的能力/solution.ts
// topics = ["二分查找"]

function shipWithinDays(weights: number[], D: number): number {
  // 判断运载能力为 w 时，能否在 D 天内运输完成
  const helper = (w: number): boolean => {
    let total = 0;
    let day = 0;
    for (const weight of weights) {
      if (total + weight > w) {
        total = weight;
        day += 1;
      } else {
        total += weight;
      }
    }
    if (total) {
      day += 1;
    }
    return day <= D;
  };

  let i = Math.max(...weights);
  let j = weights.reduce((acc, cur) => acc + cur);
  while (i < j) {
    const mid = Math.floor((i + j) / 2);
    if (!helper(mid)) {
      i = mid + 1;
    } else {
      j = mid;
    }
  }

  return i;
}
